### PR TITLE
Support mesh optimization when using AttachMeshShapeFeature

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

## Summary

Add mesh optimization support in dartsim.

Dartsim uses `AttachMeshShapeFeature` for constructing meshes. It expects a `common::Mesh` to already be loaded and passed to it. So we need to optimize the mesh first in gz-sim (instead of doing mesh optimization inside gz-physics dartsim plugin). To support this, this PR adds logic to perform mesh optimization whenever a mesh is being loaded (e.g. via `loadMesh` [here](https://github.com/gazebosim/gz-sim/blob/8ca89f7c279400060b4fc4563442f6fa521380dc/src/systems/physics/Physics.cc#L1391) in physics system).

## Test it

You can test with the [cordless drill simplified](https://app.gazebosim.org/iche033/fuel/models/Cordless%20Drill%20Simplified) mesh.  Here's a [cordless_drill.sdf](https://gist.github.com/iche033/428fd4624393576d99045b67b78b1698) test world

Launch gz-sim with the optimized mesh

```
gz sim -v 4 cordless_drill.sdf
```

and you should see a debug msg like this:

```
[Dbg] [Util.cc:918] Optimizing mesh (convex_decomposition): <home>/.gz/fuel/fuel.gazebosim.org/iche033/models/cordless drill simplified/3/meshes/cordless_drill.stl
```

Mesh collision visualization support is in https://github.com/gazebosim/gz-sim/pull/2352. I'll refactor that PR to use the new `optimizeMesh` function introduced here once this PR gets in.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/Util.cc [hunk 1]
```diff
@@ -862,8 +865,81 @@ const common::Mesh *loadMesh(const sdf::Mesh &_meshSdf)
...
+    maxConvexHulls = 1u;
+  }
+  else if (_meshSdf.ConvexDecomposition())
+  {
+    // limit max number of convex hulls to generate
+    maxConvexHulls = _meshSdf.ConvexDecomposition()->MaxConvexHulls();
+  }
+  // Check if MeshManager contains the decomposed mesh already. If not
+  // add it to the MeshManager so we do not need to decompose it again.
+  const std::string convexMeshName =
```
Comment (ahcorde) [inline src/Util.cc:904]: include `<string>`

File: src/Util.cc [hunk 2]
```diff
@@ -862,8 +865,81 @@ const common::Mesh *loadMesh(const sdf::Mesh &_meshSdf)
...
+  auto *optimizedMesh = meshManager.MeshByName(convexMeshName);
+  if (!optimizedMesh)
+  {
+    // Merge meshes before convex decomposition
+    auto mergedMesh = gz::common::MeshManager::MergeSubMeshes(_mesh);
+    if (mergedMesh && mergedMesh->SubMeshCount() == 1u)
+    {
+      // Decompose and add mesh to MeshManager
+      auto mergedSubmesh = mergedMesh->SubMeshByIndex(0u).lock();
+      std::vector<common::SubMesh> decomposed =
```
Comment (ahcorde) [inline src/Util.cc:915]: include `<vector>`

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: src/Util.cc (hunk 2 above)
Comment (iche033) [reply src/Util.cc:917]: [a428b9e](https://github.com/gazebosim/gz-sim/pull/2417/commits/a428b9e21fb412c5ad207fac2d6c82ff2b4adf57)

File: src/Util.cc (hunk 1 above)
Comment (iche033) [reply src/Util.cc:906]: [a428b9e](https://github.com/gazebosim/gz-sim/pull/2417/commits/a428b9e21fb412c5ad207fac2d6c82ff2b4adf57)

**Review by ahcorde — APPROVED** [review]

**Pull request merged**